onnxruntime/contrib_ops/cpu/qlinear_pool.cc

```diff
@@ -0,0 +1,323 @@
+    return TensorOpCost{loop_count, loop_count, loop_count};
+  }
+
+  void operator()(std::ptrdiff_t begin, std::ptrdiff_t end) const {
+    for (int64_t c = begin; c < end; ++c) {
+      operator()(c);
+    }
+  }
+
+  void operator()(std::ptrdiff_t c) const {
+    const T8Bits* x_d = X_data + c * x_step;
+    T8Bits* y_d = Y_data + c * y_step;
+
+    for (int64_t ph = 0; ph < pooled_height; ++ph) {
+      int64_t hstart = ph * stride_h - pads[0];
+      int64_t hend = std::min(hstart + kernel_shape[0], height);
+      hstart = std::max(hstart, static_cast<int64_t>(0));
+      float Yh = PoolType::Initialize();
+      for (int64_t h = hstart; h < hend; ++h) {
+        float x_value = dequantize_value(x_d[h], x_scale,x_zero_point);
+        PoolType::Process(x_value, Yh, pool_context_);
+      }
+      if (pool_attrs_.count_include_pad) {
+        PoolType::Finalize(kernel_shape[0], Yh, pool_context_);
+      } else {
+        PoolType::Finalize(hend - hstart, Yh, pool_context_);
+      }
+      y_d[ph] = quantize_value(Yh, y_scale, y_zero_point);
+    }
+  }
+};
+
+
+template <typename T8Bits, typename PoolType>
+struct QLinearPool2DTask final {
+  const T8Bits* X_data;
+  float x_scale;
+  T8Bits x_zero_point;
+  T8Bits* Y_data;
+  float y_scale;
+  T8Bits y_zero_point;
+  int64_t x_step;
+  int64_t y_step;
+  int64_t pooled_height;
+  int64_t pooled_width;
+  int64_t stride_h;
+  int64_t stride_w;
+  int64_t height;
+  int64_t width;
+  const std::vector<int64_t>& kernel_shape;
+  const std::vector<int64_t>& pads;
+  const PoolProcessContext& pool_context_;
+  const PoolAttributes& pool_attrs_;
+
+  TensorOpCost Cost() {
+    double loop_count = static_cast<double>(pooled_height * pooled_width * kernel_shape[0] * kernel_shape[1]);
+    return TensorOpCost{loop_count, loop_count, loop_count};
+  }
+
+  void operator()(std::ptrdiff_t begin, std::ptrdiff_t end) const {
+    for (int64_t c = begin; c < end; ++c) {
+      operator()(c);
+    }
+  }
+
+  void operator()(std::ptrdiff_t c) const {
+    const T8Bits* x_d = X_data + c * x_step;
+    T8Bits* y_d = Y_data + c * y_step;
+
+    for (int64_t ph = 0; ph < pooled_height; ++ph) {
+      int64_t hstart = ph * stride_h - pads[0];
+      int64_t hend = std::min(hstart + kernel_shape[0], height);
+      hstart = std::max(hstart, static_cast<int64_t>(0));
+      for (int64_t pw = 0; pw < pooled_width; ++pw) {
+        int64_t wstart = pw * stride_w - pads[1];
+        int64_t wend = std::min(wstart + kernel_shape[1], width);
+        wstart = std::max(wstart, static_cast<int64_t>(0));
+        const int64_t pool_index = ph * pooled_width + pw;
+        float Yh = PoolType::Initialize();
+        for (int64_t h = hstart; h < hend; ++h) {
+          for (int64_t w = wstart; w < wend; ++w) {
+            const int64_t input_index = h * width + w;
+            float x_value = dequantize_value(x_d[input_index], x_scale, x_zero_point);
```

Comment:
Is it computed multiple time for different pooled_height? What if we dequantize the whole input at once, then AveragePool, and quantize it back, before we have an optimized version?

Reply:
A good quick improve with some memory. Let me add some logic here.